plonky3/src/circuit_builder.rs

```diff
@@ -250,34 +259,47 @@ impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
 
 impl<'a, T: FieldElement> BaseAir<Val> for PowdrCircuit<'a, T> {
     fn width(&self) -> usize {
-        assert_eq!(self.analyzed.constant_count(), 0);
         self.analyzed.commitment_count() + 3 * self.analyzed.publics_count()
     }
 
+    fn preprocessed_width(&self) -> usize {
+        self.analyzed.constant_count()
+    }
+
     fn preprocessed_trace(&self) -> Option<RowMajorMatrix<Val>> {
+        #[cfg(debug_assertions)]
+        {
+            self.preprocessed.clone()
+        }
+        #[cfg(not(debug_assertions))]
         unimplemented!()
     }
 }
 
-impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val>> Air<AB> for PowdrCircuit<'a, T> {
+impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val> + PairBuilder> Air<AB>
+    for PowdrCircuit<'a, T>
+{
     fn eval(&self, builder: &mut AB) {
-        let matrix = builder.main();
+        let main = builder.main();
+        let fixed = builder.preprocessed();
         let pi = builder.public_values();
         let publics = self.get_publics();
         assert_eq!(publics.len(), pi.len());
 
         // public constraints
         let pi_moved = pi.to_vec();
-        let (local, next) = (matrix.row_slice(0), matrix.row_slice(1));
+        let (local, next) = (main.row_slice(0), main.row_slice(1));
+
+        let public_offset = self.analyzed.commitment_count();
 
         publics.iter().zip(pi_moved).enumerate().for_each(
             |(index, ((_, col_id, row_id), public_value))| {
                 //set decr for each public to be row_id in the first row and decrement by 1 each row
                 let (decr, inv_decr, s, decr_next) = (
-                    local[self.analyzed.commitment_count() + 3 * index],
-                    local[self.analyzed.commitment_count() + 3 * index + 1],
-                    local[self.analyzed.commitment_count() + 3 * index + 2],
-                    next[self.analyzed.commitment_count() + 3 * index],
+                    local[public_offset + 3 * index],
+                    local[public_offset + 3 * index + 1],
+                    local[public_offset + 3 * index + 2],
```

Comment:
Now we can add one fixed column instead of 3 witness columns, right?

Reply:
Yes, different pr